### PR TITLE
Move config.Map to its own package which does not depend on any component concept

### PR DESCRIPTION
**Pull request by bogdandrutu**
Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

### CODE REVIEW

**Review by mx-psi — COMMENTED** [review]
I don't have good ideas for the name but I am in favor of moving this to its own package; I think it will allow for gradually marking parts of the codebase as stable

**Comment by bogdandrutu** [comment]
@mx-psi decided on `confmap`, I think is not generic and is a cool name.

File: CHANGELOG.md [hunk 1]
```diff
@@ -19,6 +19,19 @@
 - Deprecate `componenterror` package, move everything to `component` (#5383)
 - `pcommon.Value.NewValueBytes` is deprecated in favor of `Value.NewValueMBytes` in preparation of migration to 
   immutable slices (#5367)
+- Move `config.Map` to its own package `confmap` which does not depend on any component concept (#5237)
+  - `config.Map` -> `confmap.ConfMap`
```
Comment (codeboten) [inline CHANGELOG.md:23]: Should this be `confmap.Map`? Seems a bit easier to use, not a strong opinion here

File: confmap/provider.go [hunk 2]
```diff
@@ -33,7 +33,7 @@ import (
 //		// repeat retrieve/wait/close cycle until it is time to shut down the Collector process.
 //		// ...
 //		mapProvider.Shutdown()
```
Comment (codeboten) [inline confmap/provider.go:35]: Should the example here be updated to remove references to `mapProvider`?

File: confmap/provider.go [hunk 3]
```diff
@@ -105,7 +105,7 @@ func WithRetrievedClose(closeFunc CloseFunc) RetrievedOption {
 // * CloseFunc specifies a function to be invoked when the configuration for which it was
```
Comment (codeboten) [inline confmap/provider.go:105]: the docstring for `cfgMap` needs updating

File: confmap/confmap.go [hunk 4]
```diff
@@ -97,34 +97,34 @@ func (l *Map) Set(key string, value interface{}) {
...
 }
 
 // Merge merges the input given configuration into the existing config.
 // Note that the given map may be modified.
-func (l *Map) Merge(in *Map) error {
+func (l *ConfMap) Merge(in *ConfMap) error {
 	return l.k.Merge(in.k)
 }
 
 // Sub returns new Map instance representing a sub-config of this instance.
```
Comment (codeboten) [inline confmap/confmap.go:110]: Should say `returns new ConfMap`

File: confmap/confmap.go [hunk 5]
```diff
@@ -31,38 +31,38 @@ const (
...
+func NewFromStringMap(data map[string]interface{}) *ConfMap {
+	p := New()
 	// Cannot return error because the koanf instance is empty.
 	_ = p.k.Load(confmap.Provider(data, KeyDelimiter), nil)
 	return p
 }
 
-// Map represents the raw configuration map for the OpenTelemetry Collector.
+// ConfMap represents the raw configuration map for the OpenTelemetry Collector.
 // The config.Map can be unmarshalled into the Collector's config using the "configunmarshaler" package.
```
Comment (codeboten) [inline confmap/confmap.go:52]: docstring needs updating

**Review by codeboten — COMMENTED** [review]
Just a few minor comments/suggestion and a question.

File: CHANGELOG.md (hunk 1 above)
Comment (bogdandrutu) [reply CHANGELOG.md:23]: or `confmap.Conf`?

Comment (codeboten) [reply CHANGELOG.md:23]: `confmap.Conf`

Comment (bogdandrutu) [reply CHANGELOG.md:23]: Last idea is `conf.Map`? Seems that I have 2 votes for `confmap.Conf` so will implement all changes for that proposal.

Comment (mx-psi) [reply CHANGELOG.md:23]: I like `confmap` more because `conf` can be confused with `config`

Comment (bogdandrutu) [reply CHANGELOG.md:23]: Done.

File: confmap/provider.go (hunk 3 above)
Comment (bogdandrutu) [reply confmap/provider.go:105]: Done.

File: confmap/provider.go (hunk 2 above)
Comment (bogdandrutu) [reply confmap/provider.go]: Done.

File: confmap/confmap.go (hunk 5 above)
Comment (bogdandrutu) [reply confmap/confmap.go]: Done.

File: confmap/confmap.go (hunk 4 above)
Comment (bogdandrutu) [reply confmap/confmap.go]: Done.

**Comment by bogdandrutu** [comment]
@codeboten @mx-psi all updates done, using `confmap.Conf`

**Review by mx-psi — APPROVED** [review]

**Review by codeboten — APPROVED** [review]
👍

**Comment by codeboten** [comment]
Please resolve conflicts

**Pull request merged**